File: Java/Irrgarten/src/irrgarten/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList;

/**
 * Represents the labyrinth in the game.
 * @author mariorgzlpz
 */
public class Labyrinth {
    private static char BLOCK_CHAR = 'X'; // Character representing a block in the labyrinth
    private static char EMPTY_CHAR = '-'; // Character representing an empty space in the labyrinth
    private static char MONSTER_CHAR = 'M'; // Character representing a monster in the labyrinth
    private static char COMBAT_CHAR = 'C'; // Character representing a combat space in the labyrinth
    private static char EXIT_CHAR = 'E'; // Character representing the exit in the labyrinth
    private static int ROW = 0; // Index of the row in an array
    private static int COL = 1; // Index of the column in an array
    private int nRows; // Number of rows in the labyrinth
    private int nCols; // Number of columns in the labyrinth
    private int exitRow; // Row index of the exit
    private int exitCol; // Column index of the exit
    private Monster[][] monsters; // 2D array to store monsters in the labyrinth
    private char[][] labyrinth; // 2D array to represent the labyrinth
    private Player[][] players; // 2D array to store players in the labyrinth

    /**
     * Constructor for the Labyrinth class.
     * @param nRows Number of rows in the labyrinth.
     * @param nCols Number of columns in the labyrinth.
     * @param exitRow Row index of the exit.
     * @param exitCol Column index of the exit.
     */
    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
        this.nRows = nRows ;
        this.nCols = nCols;
        this.exitRow = exitRow;
        this.exitCol = exitCol;
        
        monsters = new Monster[nRows][nCols];
        players = new Player[nRows][nCols];
        labyrinth = new char[nRows][nCols];
        
        // Initialize the labyrinth with empty spaces
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++) {
                labyrinth[i][j] = EMPTY_CHAR;
            }
            labyrinth[exitRow][exitCol] = EXIT_CHAR;
        }
    }
    
    /**
     * Spread players randomly across the labyrinth.
     * @param players The list of players to spread.
     */
    public void spreadPlayers(ArrayList<Player> players){
        for (Player p : players) {
          int[] pos = randomEmptyPos();
          int oldRow = -1 ,oldCol = -1;
          Monster monster = putPlayer2D(oldRow,oldCol,pos[ROW],pos[COL],p);
        }
    }
    
    /**
     * Checks if there's a winner in the game.
     * @return True if there's a winner, false otherwise.
     */
    public boolean haveAWinner(){
        return players[exitRow][exitCol] != null;
    }
    
    /**
     * Gets the string representation of the labyrinth.
     * @return String representation of the labyrinth.
     */
    public String toString() {
        String result = "";

        // Iterate over each row of the labyrinth
        for (int i = 0; i < nRows; i++) {
            // Iterate over each column of the labyrinth in the current row
            for (int j = 0; j < nCols; j++) {
                // Add the character at the current position to the resulting string
                result += labyrinth[i][j] + "   ";
            }
            // Add a newline at the end of each row
            result += "\n";
        }

        // Return the resulting string
        return result;
    }
    
    /**
     * Adds a monster to the specified position in the labyrinth.
     * @param row The row index of the position.
     * @param col The column index of the position.
     * @param monster The monster to add.
     */
    public void addMonster(int row, int col, Monster monster){
        if(this.emptyPos(row, col) && this.posOK(row, col)){
            labyrinth[row][col] = 'M';
            monsters[row][col] = monster;
            monster.setPos(row, col);
        }
    }
    
    /**
     * Moves a player in the specified direction and handles combat if necessary.
     * @param direction The direction in which to move the player.
     * @param player The player to move.
     * @return The monster involved in the combat if any, null otherwise.
     */
    public Monster putPlayer(Directions direction, Player player){
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        int[] newPos = dir2Pos(oldRow,oldCol,direction);
        Monster monster = putPlayer2D(oldRow,oldCol,newPos[0],newPos[1],player);
        return monster;
    }
    
    /**
     * Adds a block to the labyrinth.
     * @param orientation The orientation of the block (HORIZONTAL or VERTICAL).
     * @param startRow The starting row index of the block.
     * @param startCol The starting column index of the block.
     * @param length The length of the block.
     */
    public void addBlock(Orientation orientation, int startRow, int startCol, int length){
        int incRow, incCol;
        if(orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        }
        else{
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow;
        int col = startCol;
        
        while((posOK(row,col) && emptyPos(row,col)) && length > 0){
            labyrinth[row][col] = BLOCK_CHAR;
            row += incRow;
            col += incCol;
            length -=1;
            
        }
    }
    
    /**
     * Gets the valid moves for a player at the specified position in the labyrinth.
     * @param row The row index of the player's position.
     * @param col The column index of the player's position.
     * @return The list of valid moves.
     */
    public ArrayList<Directions> validMoves(int row, int col){
        ArrayList<Directions> output = new ArrayList<>();

        if (canStepOn(row + 1, col)) {
            output.add(Directions.DOWN);
        }
        if (canStepOn(row - 1, col)) {
            output.add(Directions.UP);
        }
        if (canStepOn(row, col + 1)) {
            output.add(Directions.RIGHT);
        }
        if (canStepOn(row, col - 1)) {
            output.add(Directions.LEFT);
        }

        return output;
    }
    
    /**
     * Checks if the given position is within the bounds of the labyrinth.
     * @param row The row index.
     * @param col The column index.
     * @return True if the position is within bounds, false otherwise.
     */
    private boolean posOK(int row, int col){
        return (row >= 0 && row < nRows && col >= 0 && col < nCols);       
    }
    
    /**
     * Checks if the specified position in the labyrinth is empty.
     * @param row The row index.
     * @param col The column index.
     * @return True if the position is empty, false otherwise.
     */
    private boolean emptyPos(int row, int col){
        return labyrinth[row][col] == '-';
    }
    
    /**
     * Checks if the specified position in the labyrinth contains a monster.
     * @param row The row index.
     * @param col The column index.
     * @return True if the position contains a monster, false otherwise.
     */
    private boolean monsterPos(int row, int col){
        return labyrinth[row][col] == 'M';
    }
    
    /**
     * Checks if the specified position in the labyrinth is the exit.
     * @param row The row index.
     * @param col The column index.
     * @return True if the position is the exit, false otherwise.
     */
    private boolean exitPos(int row, int col){
        return labyrinth[row][col] == 'E';  
    }
    
    /**
     * Checks if the specified position in the labyrinth is a combat space.
     * @param row The row index.
     * @param col The column index.
     * @return True if the position is a combat space, false otherwise.
     */
    private boolean combatPos(int row, int col){
        return labyrinth[row][col] == 'C';
    }
    
    /**
     * Checks if a player can step on the specified position in the labyrinth.
     * @param row The row index.
     * @param col The column index.
     * @return True if the player can step on the position, false otherwise.
     */
    private boolean canStepOn(int row, int col){
        return this.posOK(row, col) && (this.emptyPos(row, col) || this.monsterPos(row, col) || this.exitPos(row, col));
    }
    
    /**
     * Updates the old position of a player or monster after moving.
     * @param row The row index of the old position.
     * @param col The column index of the old position.
     */
    private void updateOldPos(int row, int col){
        if(this.posOK(row, col)){
            if(this.combatPos(row, col)){
                labyrinth[row][col] = 'M';
            }
            else{
                labyrinth[row][col] = '-';
            }    
        }
    }
    
    /**
     * Converts a direction to a new position.
     * @param row The current row index.
     * @param col The current column index.
     * @param direction The direction to move.
     * @return An array containing the new row and column indices.
     */
    private int[] dir2Pos(int row, int col, Directions direction){
        switch(direction){
            case LEFT:
                col -= 1;
                break;
            case RIGHT:
                col += 1;
                break;
            case UP:
                row -= 1;
                break;
            case DOWN:
                row += 1; 
                break;
        }
        int[] posicion = new int[]{row, col};
        return posicion;
    }
    
    /**
     * Finds a random empty position in the labyrinth.
     * @return An array containing the row and column indices of the empty position.
     */
    private int[] randomEmptyPos(){
        int row = Dice.randomPos(nRows);
        int col = Dice.randomPos(nCols);
        while(!this.emptyPos(row, col)){
            row = Dice.randomPos(nRows);
            col = Dice.randomPos(nCols);
        }
        return new int[]{row, col};
    }
    
    /**
     * Places a player in the specified position in the labyrinth and handles combat.
     * @param oldRow The old row index of the player.
     * @param oldCol The old column index of the player.
     * @param row The new row index of the player.
     * @param col The new column index of the player.
     * @param player The player to place.
     * @return The monster involved in combat, if any.
     */
    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player){
        Monster output = null;
        if(canStepOn(row,col)){
            if(posOK(oldRow,oldCol)){
                Player p = players[oldRow][oldCol];
                if(p == player){
                    updateOldPos(oldRow,oldCol);
                    players[oldRow][oldCol] = null;
                }
            }
            boolean monsterPos = monsterPos(row,col);
            if(monsterPos){
                labyrinth[row][col] = COMBAT_CHAR;
                output = monsters[row][col];
            }
            else {
                char number = player.getNumber();
                labyrinth[row][col] = number;
            }
            
            players[row][col] = player;
            player.setPos(row, col);
        }
        return output;
    } 
}

File: Ruby/Labyrinth.rb
#encoding utf-8

require_relative 'Monster'
require_relative 'Player'
require_relative 'Directions'

module Irrgarten
    class Labyrinth

        @@BLOCK_CHAR = 'X'
        @@EMPTY_CHAR = '-'
        @@MONSTER_CHAR = 'M'
        @@COMBAT_CHAR = 'C'
        @@EXIT_CHAR = 'E'
        @@ROW = 0
        @@COL = 1

        def initialize(a_n_rows, a_n_cols, a_exit_row, a_exit_col)
            @n_rows = a_n_rows
            @n_cols = a_n_cols
            @exit_row = a_exit_row
            @exit_col = a_exit_col
            @monsters = Array.new(@n_rows) { Array.new(@n_cols, nil) }
            @players = Array.new(@n_rows) { Array.new(@n_cols, nil) }
            @labyrinth = Array.new(@n_rows) { Array.new(@n_cols, @@EMPTY_CHAR) }
        
            @n_rows.times do |i|
              @n_cols.times do |j|
                @labyrinth[i][j] = @@EMPTY_CHAR
              end
            end
            @labyrinth[@exit_row][@exit_col] = @@EXIT_CHAR
        end

        def set_square(row, col, value)
            @labyrinth[row][col] = value
        end

        def set_fuzzy_player(row, col, fuzzy)
            @players[row][col] = fuzzy
        end


        def spread_players(players)
            players.each do |player|
                pos = random_empty_pos
                old_row = -1
                old_col = -1
                monster = put_player_2D(old_row,old_col,pos[@@ROW], pos[@@COL], player)
            end
        end
    
        def have_a_winner
            return @players[@exit_row][@exit_col] != nil
        end
    
        def to_s
        result = ""
    
        @n_rows.times do |i|
            @n_cols.times do |j|
            result += "#{@labyrinth[i][j]}   "
            end
            result += "\n"
        end
    
        result
        end
        
        def add_monster(row, col, monster)
            if empty_pos(row, col) && pos_ok(row, col)
                @labyrinth[row][col] = @@MONSTER_CHAR
                @monsters[row][col] = monster
                monster.set_pos(row, col)
            end
        end
    
        def put_player(direction, player)
            old_row = player.row
            old_col = player.col
            new_pos = dir_2_pos(player.row, player.col, direction)
            monster = put_player_2D(old_row, old_col, new_pos[@@ROW], new_pos[@@COL], player)
            return monster
        end
    
        def add_block(orientation, start_row, start_col, length)
            if (orientation == Orientation::HORIZONTAL)
                inc_row = 0
                inc_col = 1 
            else
                inc_row = 1
                inc_col = 0 
            end
            row = start_row
            col = start_col
            length.times do
                if (pos_ok(row, col) && empty_pos(row, col))
                    @labyrinth[row][col] = @@BLOCK_CHAR
                end
                row += inc_row
                col += inc_col
            end
        end
    
        def valid_moves(row, col)
            output = []
            output << Directions::UP if can_step_on(row - 1, col)
            output << Directions::DOWN if can_step_on(row + 1, col)
            output << Directions::LEFT if can_step_on(row, col - 1)
            output << Directions::RIGHT if can_step_on(row, col + 1)
            return output

        end
    
        def pos_ok(row, col)
            return row >= 0 && row < @n_rows && col >= 0 && col < @n_cols
        end
    
        def empty_pos(row, col)
            return @labyrinth[row][col] == @@EMPTY_CHAR
        end
    
        def monster_pos(row, col)
            return @labyrinth[row][col] == @@MONSTER_CHAR
        end
    
        def exit_pos(row, col)
            return @labyrinth[row][col] == @@EXIT_CHAR
        end
    
        def combat_pos(row, col)
            return @labyrinth[row][col] == @@COMBAT_CHAR
        end
    
        def can_step_on(row, col)
            pos_ok(row, col) && (empty_pos(row, col) || monster_pos(row, col) || exit_pos(row, col))
        end
    
        def update_old_pos(row, col)
            if pos_ok(row, col)
                if combat_pos(row, col)
                    @labyrinth[row][col] = @@MONSTER_CHAR
                else
                    @labyrinth[row][col] = @@EMPTY_CHAR
                end
            end
        end
    
        def dir_2_pos(row, col, direction)
            case direction
            when Directions::LEFT
                col -= 1
            when Directions::RIGHT
                col += 1
            when Directions::UP
                row -= 1
            when Directions::DOWN
                row += 1
            end
            return [row, col]
        end
    
        def random_empty_pos
            row = Dice.random_pos(@n_rows)
            col = Dice.random_pos(@n_cols)
            while(!empty_pos(row,col))
                row = Dice.random_pos(@n_rows)
                col = Dice.random_pos(@n_cols)
            end
            return [row, col]
        end
    
        def put_player_2D(old_row, old_col, row, col, player)
            output = nil
            if (can_step_on(row, col))
                if pos_ok(old_row, old_col)
                    jugador = @players[old_row][old_col]
                    if(jugador == player)
                        update_old_pos(old_row, old_col)
                        @players[old_row][old_col] = nil
                    end
                end
                monster_p = monster_pos(row, col)
                if monster_p
                    @labyrinth[row][col] = @@COMBAT_CHAR
                    output = @monsters[row][col]
                else
                    number = player.number
                    @labyrinth[row][col] = number.to_s
                end
                @players[row][col] = player
                player.set_pos(row, col)
            end
            return output
        end

        private :pos_ok, :empty_pos, :monster_pos, :exit_pos, :combat_pos, :can_step_on, :update_old_pos, :dir_2_pos, :random_empty_pos, :put_player_2D
    end
end